cpp/include/patterns/vertex_frontier.cuh

```diff
@@ -37,129 +36,74 @@
-  size_t bucket_block_aggregate_sizes[num_buckets];
-
-  while (block_idx < num_blocks) {
-    for (size_t i = 0; i < num_buckets; ++i) { bucket_block_local_offsets[i] = 0; }
-
-    size_t selected_bucket_idx{invalid_bucket_idx};
-    vertex_t key{invalid_vertex};
-
-    if (idx < num_elements) {
-      key                 = *(row_first + idx);
-      selected_bucket_idx = split_op(key);
-      if (selected_bucket_idx != this_bucket_idx) {
-        *(row_first + idx) = invalid_vertex;
-        if (selected_bucket_idx != invalid_bucket_idx) {
-          bucket_block_local_offsets[selected_bucket_idx] = 1;
-        }
-      }
-    }
-
-    for (size_t i = 0; i < num_buckets; ++i) {
-      BlockScan(temp_storage)
-        .ExclusiveSum(bucket_block_local_offsets[i],
-                      bucket_block_local_offsets[i],
-                      bucket_block_aggregate_sizes[i]);
-    }
-
-    if (threadIdx.x == 0) {
-      for (size_t i = 0; i < num_buckets; ++i) {
-        static_assert(sizeof(unsigned long long int) == sizeof(size_t));
-        bucket_block_start_offsets[i] =
-          atomicAdd(reinterpret_cast<unsigned long long int*>(bucket_sizes_ptr + i),
-                    static_cast<unsigned long long int>(bucket_block_aggregate_sizes[i]));
-      }
-    }
-
-    __syncthreads();
-
-    // FIXME: better use shared memory buffer to aggreaget global memory writes
-    if ((selected_bucket_idx != this_bucket_idx) && (selected_bucket_idx != invalid_bucket_idx)) {
-      bucket_ptrs[selected_bucket_idx][bucket_block_start_offsets[selected_bucket_idx] +
-                                       bucket_block_local_offsets[selected_bucket_idx]] = key;
-    }
-
-    idx += gridDim.x * blockDim.x;
-    block_idx += gridDim.x;
-  }
-}
-
-}  // namespace detail
-
 template <typename vertex_t, bool is_multi_gpu = false>
-class Bucket {
+class SortedUniqueElementBucket {
  public:
-  Bucket(raft::handle_t const& handle, size_t capacity)
-    : handle_ptr_(&handle), elements_(capacity, handle.get_stream())
+  SortedUniqueElementBucket(raft::handle_t const& handle)
+    : handle_ptr_(&handle), elements_(0, handle.get_stream())
   {
-    thrust::fill(rmm::exec_policy(handle_ptr_->get_stream())->on(handle_ptr_->get_stream()),
-                 elements_.begin(),
-                 elements_.end(),
-                 invalid_vertex_id<vertex_t>::value);
   }
 
   void insert(vertex_t v)
   {
-    raft::update_device(elements_.data() + size_, &v, 1, handle_ptr_->get_stream());
-    ++size_;
+    if (elements_.size() > 0) {
+      rmm::device_scalar<vertex_t> vertex(v, handle_ptr_->get_stream());
+      insert(vertex.data(), vertex_t{1});
+    } else {
+      elements_.resize(1, handle_ptr_->get_stream());
+      raft::update_device(elements_.data(), &v, size_t{1}, handle_ptr_->get_stream());
+    }
   }
 
-  size_t size() const { return size_; }
+  void insert(vertex_t const* sorted_unique_vertices, vertex_t num_sorted_unique_vertices)
```

Comment:
`sorted_unique_vertices` is assumed to be device memory right? Consider mentioning it or having a check for this.
It could be confusing because the other `insert` function takes host memory.

Reply:
insert(vertex_t v) takes a scalar and otherwise explicitly mentioned, all pointers are device pointers. I will add a comment to additionally mark this requires a device memory pointer.

Eventually, I think we should adopt device_span and host_span from cuDF to avoid this kind of confusion.